### PR TITLE
Added an attribute to remove non-codegen components

### PR DESCRIPTION
**Pull request by samiwh**
#### Description
Currently we hard code in temporary components in the clean system. This makes this generic and opens up the possibility of having auto cleaned components in other features.
#### Tests
There is a clean reactive component test already, which this passes
#### Documentation
temporary-component.md
#### Primary reviewers
@zeroZshadow @jamiebrynes7 @JonasImprobable

### CODE REVIEW

File: workers/unity/Assets/Gdk/Core/Systems/CleanReactiveComponentsSystem.cs [hunk 1]
```diff
@@ -19,53 +26,76 @@ protected override void OnCreateManager(int capacity)
...
-            onConnectedComponentGroup = GetComponentGroup(typeof(OnConnected));
-            onDisconnectedComponentGroup = GetComponentGroup(typeof(OnDisconnected));
         }
 
         private void GenerateComponentGroups()
         {
             foreach (var translationUnit in view.TranslationUnits.Values)
             {
-                translationUnit.CleanUpComponentGroups = new List<ComponentGroup>();
+                translationUnit.CleanUpComponentGroups = new Collections.List<ComponentGroup>();
```
Comment (jamiebrynes7) [inline workers/unity/Assets/Gdk/Core/Systems/CleanReactiveComponentsSystem.cs]: Unnecessary change?

File: workers/unity/Assets/Gdk/Core/Systems/CleanReactiveComponentsSystem.cs [hunk 2]
```diff
@@ -19,53 +26,76 @@ protected override void OnCreateManager(int capacity)
...
             }
+
+            MethodInfo addRemoveComponentActionMethod =
+                typeof(CleanReactiveComponentsSystem).GetMethod("AddRemoveComponentAction",
+                    BindingFlags.NonPublic | BindingFlags.Instance);
+
+            // Find all components with the RemoveAtEndOfTick attribute
+            foreach (var assembly in AppDomain.CurrentDomain.GetAssemblies())
+            {
+                foreach (var type in assembly.GetTypes())
```
Comment (jamiebrynes7) [inline workers/unity/Assets/Gdk/Core/Systems/CleanReactiveComponentsSystem.cs]: LINQ ?

`assembly.GetTypes().Where((type) => ...).ForEach((type) => ...)`

File: workers/unity/Assets/Gdk/Core/Systems/CleanReactiveComponentsSystem.cs (hunk 1 above)
Comment (samiwh) [reply workers/unity/Assets/Gdk/Core/Systems/CleanReactiveComponentsSystem.cs]: Accident. Will fix.

File: workers/unity/Assets/Gdk/Core/Systems/CleanReactiveComponentsSystem.cs (hunk 2 above)
Comment (samiwh) [reply workers/unity/Assets/Gdk/Core/Systems/CleanReactiveComponentsSystem.cs]: I personally think that makes this harder to read. But I will see if there is something I can do with this.

File: workers/unity/Assets/Gdk/Core/Systems/CleanReactiveComponentsSystem.cs [hunk 3]
```diff
@@ -1,15 +1,22 @@
-using Improbable.Collections;
+using System;
+using System.Collections.Generic;
+using System.Reflection;
+using JetBrains.Annotations;
```
[the remark this answers is not in the document]
Comment (samiwh) [reply workers/unity/Assets/Gdk/Core/Systems/CleanReactiveComponentsSystem.cs]: It's for the [UsedImplicitly] annotation. That stops resharper/rider from complaining. I didn't realise it had added an include. Will take a look at whether this will compile on vanilla vs and see if I can get rid of the warning another way.

File: docs/content/temporary-components.md [hunk 4]
```diff
@@ -0,0 +1,43 @@
+**Warning:** The [pre-alpha](https://docs.improbable.io/reference/latest/shared/release-policy#maturity-stages) release is for evaluation purposes only, with limited documentation - see the guidance on [Recommended use](../../README.md#recommended-use).
+
+-----
+
+
+## Temporary components
+
+The attribute `Improbable.Gdk.Core.RemoveAtEndOfTick`, can be applied to any ECS component, extending either `IComponentData` or `ISharedComponentData`.
```
Comment (firtina-improbable) [inline docs/content/temporary-components.md]: `Tick` > `Update` perhaps?

File: workers/unity/Assets/Gdk/Core/Components/WorkerEntityComponents.cs [hunk 5]
```diff
@@ -2,20 +2,40 @@
...
+    /// Component added to the worker entity immediately after establishing a connection to a SpatialOS deployment
+    /// Removed immediately after disconnecting
+    /// </summary>
     public struct IsConnected : IComponentData
     {
     }
 
+    /// <summary>
+    /// Component added to the worker entity immediately after establishing a connection to a SpatialOS deployment
+    /// Removed at the end of the tick is was added
```
Comment (firtina-improbable) [inline workers/unity/Assets/Gdk/Core/Components/WorkerEntityComponents.cs]: "Removed at the end of the tick is was added" sounds odd

File: workers/unity/Assets/Gdk/Core/Systems/CleanReactiveComponentsSystem.cs [hunk 6]
```diff
@@ -35,37 +38,64 @@ private void GenerateComponentGroups()
...
+                typeof(CleanReactiveComponentsSystem).GetMethod("AddRemoveComponentAction",
+                    BindingFlags.NonPublic | BindingFlags.Instance);
+
+            // Find all components with the RemoveAtEndOfTick attribute
+            foreach (var assembly in AppDomain.CurrentDomain.GetAssemblies())
+            {
+                foreach (var type in assembly.GetTypes())
+                {
+                    if (type.GetCustomAttribute<RemoveAtEndOfTick>(false) == null
+                        || typesToRemove.Contains(type)
```
Comment (firtina-improbable) [inline workers/unity/Assets/Gdk/Core/Systems/CleanReactiveComponentsSystem.cs]: I recommend moving `typesToRemove.Contains` check at the end of this conditional block, or even after it as a separate `continue`. This is because it's currently between two type-based conditions

File: docs/content/temporary-components.md [hunk 7]
```diff
@@ -0,0 +1,43 @@
+**Warning:** The [pre-alpha](https://docs.improbable.io/reference/latest/shared/release-policy#maturity-stages) release is for evaluation purposes only, with limited documentation - see the guidance on [Recommended use](../../README.md#recommended-use).
```
Comment (jessicafalk) [inline docs/content/temporary-components.md]: @ElleEss should go through the doc before merging it in.

File: docs/content/temporary-components.md [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+-----
+
+
+## Temporary components
```
Comment (jessicafalk) [inline docs/content/temporary-components.md]: Still think that some kind of motivation as to why we need temporary components is missing here. The doc just jumps straight to the implementation

File: docs/content/temporary-components.md [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+The attribute `Improbable.Gdk.Core.RemoveAtEndOfTick`, can be applied to any ECS component, extending either `IComponentData` or `ISharedComponentData`.
+All components with this attribute will be removed from all entities during the `InternalSpatialOSCleanGroup` at the end of `SpatialOSSendGroup`.
```
Comment (jessicafalk) [inline docs/content/temporary-components.md]: do we have a ticket to update the doc after we added the documentation about the groups?

File: docs/content/temporary-components.md [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+    [Inject] Data data;
+
+    protected override void OnUpdate()
+    {
+        // perform work related to the temporary component
+    }
+}
+``` 
+
+At the end of the update loop, `SomeTemporaryComponent` will automatically be removed from all entities.
```
Comment (jessicafalk) [inline docs/content/temporary-components.md]: should we mention here during which system update that happens

File: docs/content/temporary-components.md (hunk 9 above)
Comment (jamiebrynes7) [reply docs/content/temporary-components.md]: I think this is consistent with the current implementation of update groups. We just haven't written the doc explaining exactly what these are yet 😬

Comment (jessicafalk) [reply docs/content/temporary-components.md]: it's more about adding a reference to that doc as soon as it exists ;)

File: docs/content/temporary-components.md (hunk 4 above)
Comment (samiwh) [reply docs/content/temporary-components.md]: It's removed at the end of the update look but it's not at the end of update. Also that is an implementation detail (although I'd be surprised if we made it so that SpatialOS ticks are not tied to update)
The intention was to show that it is based around the SpatialOS tick.

File: workers/unity/Assets/Gdk/Core/Systems/CleanReactiveComponentsSystem.cs (hunk 6 above)
Comment (samiwh) [reply workers/unity/Assets/Gdk/Core/Systems/CleanReactiveComponentsSystem.cs]: Will do

File: docs/content/temporary-components.md (hunk 7 above)
Comment (samiwh) [reply docs/content/temporary-components.md]: Will add as a reviewer.

File: docs/content/temporary-components.md (hunk 8 above)
Comment (samiwh) [reply docs/content/temporary-components.md]: That's mostly because we are missing all the things that go with it and the three or so docs that also need to be included. This is factored out from other stuff.
The reason for it to be there is mostly the temporary components on the worker entity and also a prerequisite for the timer systems and various other things.

File: docs/content/temporary-components.md (hunk 9 above)
Comment (samiwh) [reply docs/content/temporary-components.md]: There is a ticket for that doc to exist.
There is another one to reorganise the docs completely too.

File: docs/content/temporary-components.md (hunk 10 above)
Comment (samiwh) [reply docs/content/temporary-components.md]: I'll mention the group. The system itself is an implementation detail. It's also easier to miss that when renaming and moving things and leave the docs out of date.

File: workers/unity/Assets/Gdk/Core/Components/WorkerEntityComponents.cs (hunk 5 above)
Comment (samiwh) [reply workers/unity/Assets/Gdk/Core/Components/WorkerEntityComponents.cs]: s/is/it

File: docs/content/temporary-components.md (hunk 4 above)
Comment (firtina-improbable) [reply docs/content/temporary-components.md]: We may need to link to what a `tick` is in that case, and what "being at the end of it" means specifically

**Comment by jamiebrynes7** [comment]
@samiwh Please add a unit test for this feature. :)

File: workers/unity/Assets/Gdk/Core/Attributes/RemoveAtEndOfTick.cs [hunk 11]
```diff
@@ -0,0 +1,15 @@
+﻿using System;
+using Unity.Entities;
+
+namespace Improbable.Gdk.Core
+{
+    /// <summary>
+    /// Any component with this attribute will be removed from all entities by the CleanReactiveComponentSystem
+    /// Can be added to components extending either <see cref="IComponentData" /> or <see cref="ISharedComponentData" />
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Struct | AttributeTargets.Class, AllowMultiple = false, Inherited = false)]
```
Comment (zeroZshadow) [inline workers/unity/Assets/Gdk/Core/Attributes/RemoveAtEndOfTick.cs]: Just wondering, what kind of classes will this be on?

Comment (samiwh) [reply workers/unity/Assets/Gdk/Core/Attributes/RemoveAtEndOfTick.cs]: I think that is a mistake. Will remove it. (Could conceivably add it to components but that will break if done right now)

File: docs/content/temporary-components.md (hunk 4 above)
Comment (samiwh) [reply docs/content/temporary-components.md]: We should, but there isn't a doc for that yet. It will likely be in the system execution order doc.

File: workers/unity/Assets/Gdk/Core/Components/WorkerEntityComponents.cs [hunk 12]
```diff
@@ -1,21 +1,58 @@
...
+    ///     Component added to the worker entity immediately after establishing a connection to a SpatialOS deployment
+    ///     Removed at the end of the tick it was added
+    /// </summary>
+    [RemoveAtEndOfTick]
     public struct OnConnected : IComponentData
     {
+        private var x = MyCoponent.SomeList;
     }
 
+    public struct MyComponent : IComponentData
```
Comment (firtina-improbable) [inline workers/unity/Assets/Gdk/Core/Components/WorkerEntityComponents.cs]: Is this block supposed to contain these lines?

Comment (jamiebrynes7) [inline workers/unity/Assets/Gdk/Core/Components/WorkerEntityComponents.cs]: What is this lol
